### ISSUE TITLE
[feature request] Pre-backward hooks for runtime profiling use case

### ISSUE THREAD

**Issue by ShreyaR** [pytorch/pytorch#42824]
Would it be possible to get a pre-backward hook, in addition to pre-forward, forward and pre-backward?

**My use case and current solution**
I want to profile the per-layer runtime of my model to identify the slowest layers. How I currently do this:

1. Assign a unique name to every leaf layer.
2. Register forward and backward hooks on every leaf layer of the model.
3. Torch.cuda.synchronize() and log the timestamp at which the hook for each layer is called.
4. Take the difference between subsequent timestamps in the log.

This gives me accurate per-layer computation time. However, the torch.cuda.synchronize() after each layer does not reflect the actual time the forward and backward prop takes.

**What I'd like to do instead**
I'd like to use torch.cuda.Events instead. This is easy to do fr the forward prop:
1. Have a start event in the pre-forward hook for each layer.
2. Have an end event in the forward hook for each layer.
3. Measure the elapsed time between the 2 events.

However, I can't do this for backprop because there isn't a pre-backward hook. Having access to pre-backward hooks will enable accurate runtime profiling for backprop.

Also, please let me know if there are alternatives to get accurate layer-wise runtime profiling using any other methods that don't have a reliance on a pre-backward hook.

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @robieta @chaekit @aaronenyeshi @ngimel @nbcsm @guotuofeng @guyang3532 @gaoteng-git @tiffzhaofb @ilia-cher @gdankel @bitfort @SsnL @gqchen

**Comment by zou3519** [comment]
Have you tried to use [`torch.autograd.profiler`](https://pytorch.org/docs/stable/autograd.html#torch.autograd.profiler.profile)?

**Comment by albanD** [comment]
Hi,

I don't think you should be the backward hooks for this. As mentioned in the doc, they are kind of broken right now and won't give you any proper timing.
You will have to put hooks in the input/output Tensors (for example during the forward hook) to get this information accurately.

**Comment by godelshana** [comment]
are there alternatives to get accurate layer-wise runtime profiling？it is import to locate the bottleneck，and speed up the module by wiriting cuda kernels。

**Comment by albanD** [comment]
Now that we have working backward hooks, we could definitely extend them to also have `pre_full_backward_hook`.

**Comment by MicPie** [comment]
@albanD Do you have more information on how to approach a possible backward pre hook setup?
I'm curious because I would need a backward pre hook setup for a project.
(We can also shift the discussion to the PyTorch forums to not derail this issue, if this is preferred. I have the same username there.)

**Comment by albanD** [comment]
Here is good to discuss new feature implementation.

At a high level, we need to update the nn.Module class to have a new pre_hook method and store these properly (just like is done for the back hook).

The in the call_impl function, we can update the code (especially the part that attach to the outputs) to pass in these new pre-hook.

At the lowest level, this function:  https://github.com/pytorch/pytorch/blob/019259a66fe6f3bd8405c5801e7d623d83a64493/torch/utils/hooks.py#L166 is already called when the output gradients are computed (to store the grad_outputs for the post hook that is called later).
You can even see there that in the case where we know inputs don't require gradient, we directly call the user post hook here.
We can extend this function to also call a user-provided pre-hook (and store and return the updated value).

**Comment by collinmccarthy** [comment]
I'm also interested in this feature, any progress? I use CUDA-based timers for a few modules in the forward pass and I'm looking for an easy way to add them for the backward pass (only timing these modules). I don't want to invoke the profiler for this (it's going to be measured during training, and if done correctly CUDA timers have almost zero overhead).

**Comment by albanD** [comment]
Increasing priority due to user activity.
